### ISSUE TITLE
Incorrect Placement of optimizer.zero_grad()

### ISSUE THREAD

**Issue by ElFatemehHonarvar** [SharifiZarchi/Introduction_to_Machine_Learning#55]
In the training loop, the optimizer.zero_grad() function is called after optimizer.step(). This order causes gradients from the current batch to remain in memory and accumulate with the gradients of subsequent batches during training.
